File: tsl/tsl/report/lab_report/lab_report.py
# Copyright (c) 2024, Tsl and contributors
# For license information, please see license.txt


import frappe
from frappe import _
from datetime import datetime,timedelta



def execute(filters=None):
	columns=get_columns(filters)
	data = get_data(filters) 
	return columns, data

def get_columns(filters):

	columns = [
		_("Work Order") + ":Link/Work Order Data:150",
		_("Purchase Order") + ":Link/Purchase Order:150",
		_("Purchase Invoice") + ":Link/Purchase Invoice:150",
		_("Tracking ID") + ":Data:150",
		_("Customer") + ":Data:200",
		_("Sales Person") + ":Data:150",
		_("Received Date") + ":Date:150",
		_("Repaired Date") + ":Date:150",
		_("Delivery Date") + ":Date:150",
		_("Status") + ":Data:150",
		_("Evaluation Report Status") + ":Data:200",
		_("Model Name") + ":Data:150",
		_("Technical Remarks") + ":Data:150",
		
	]	
	return columns


def get_data(filters):
	data = []
	if filters.work_order_data:
		evr = frappe.get_all("Evaluation Report",{"work_order_data":filters.work_order_data ,"docstatus":1,"date": ["between", [filters.from_date,filters.to_date]]},["*"])
	
	elif filters.branch:
		evr = frappe.get_all("Evaluation Report",{"branch":filters.branch ,"docstatus":1,"date": ["between", [filters.from_date,filters.to_date]]},["*"])
	
	else:
		evr = frappe.get_all("Evaluation Report",{"docstatus":1,"date": ["between", [filters.from_date,filters.to_date]]},["*"])

	for i in evr:
		rec_date = frappe.get_value("Work Order Data",{"name":i.work_order_data},["posting_date"])
		wods= frappe.get_value("Work Order Data",{"name":i.work_order_data},["name"])
		cus = frappe.get_value("Work Order Data",{"name":i.work_order_data},["customer"])
		dn_date = frappe.get_value("Work Order Data",{"name":i.work_order_data},["dn_date"])
		w_status= frappe.get_value("Work Order Data",{"name":i.work_order_data},["status"])
		sales= frappe.get_value("Work Order Data",{"name":i.work_order_data},["sales_rep"])


		psi = frappe.db.sql(""" select `tabEvaluation Report`.comment as com,`tabEvaluation Report`.test_result as ts,`tabPart Sheet Item`.model  from `tabEvaluation Report` 
		left join `tabPart Sheet Item` on `tabEvaluation Report`.name = `tabPart Sheet Item`.parent
		where `tabEvaluation Report`.name = %s 
		""",i.name,as_dict=1)
		
		if psi:
			for j in psi:
				im = frappe.get_value("Item Model",{"name":j["model"]},["model"])

				po = ""
				p_order = frappe.db.sql(""" select `tabPurchase Order`.name as po from `tabPurchase Order` 
				left join `tabPurchase Order Item` on `tabPurchase Order`.name = `tabPurchase Order Item`.parent 
				where `tabPurchase Order Item`.work_order_data = '%s' """ %(wods),as_dict =1)
				if p_order:
					po = p_order[0]["po"]
			
				pi = ""
				tr_id = ""
						
				p_inv = frappe.db.sql(""" select `tabPurchase Invoice`.tracking_id as td,`tabPurchase Invoice`.name as pi from `tabPurchase Invoice` 
				left join `tabPurchase Invoice Item` on `tabPurchase Invoice`.name = `tabPurchase Invoice Item`.parent 
				where `tabPurchase Invoice Item`.work_order_data = '%s' """ %(wods),as_dict =1)
				if p_inv:
					pi = p_inv[0]["pi"]
					tr_id = p_inv[0]["td"]

								
				row = [wods,po,pi,tr_id,cus,sales,rec_date,i.date,dn_date or "-",w_status,i.status,im,j["com"] or j["ts"]]
				data.append(row)
			data.append(["","","","","","","","","","","","",""])		
	return data

	

File: tsl/tsl/report/lab_report/lab_report.js
// Copyright (c) 2024, Tsl and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Lab Report"] = {
	"filters": [

		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"width": "80",
			"reqd": 1,
			default: frappe.datetime.add_months(frappe.datetime.month_start(), -1),
			
	
		},
		
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1
			
		},

		{
			"fieldname": "work_order_data",
			"label": __("Work Order"),
			"fieldtype": "Link",
			"width": "80",
			"options": "Work Order Data",

			
		},


		{
			"fieldname": "branch",
			"label": __("Branch"),
			"fieldtype": "Link",
			"width": "80",
			"options": "Branch",

			
		},


		// {
		// 	"fieldname": "tracking_id",
		// 	"label": __("Tracking ID"),
		// 	"fieldtype": "Data",
		// 	"width": "80",
			

			
		// },


		
		// {
		// 	"fieldname": "status",
		// 	"label": __("Status"),
		// 	"fieldtype": "Select",
		// 	"options": "A-Approved\nAP-Available Parts\nC-Comparison\nCT-Customer Testing\nCC-Comparison Client\nEP-Extra Parts\nIQ-Internally Quoted\nNE-Need Evaluation\nNER-Need Evaluation Return\nNEA-Need Evaluation Approved\nP-Paid\nQ-Quoted\nRNA-Return Not Approved\nRNAC-Return Not Approved Client\nRNF-Return No Fault\nRNFC-Return No Fault Client\nRNP-Return No Parts\nRNPC-Return No Parts Client\nRNR-Return Not Repaired\nRNRC-Return Not Repaired Client\nRS-Repaired and Shipped\nRSC-Repaired and Shipped Client\nRSI-Repaired and Shipped Invoiced\nSP-Searching Parts\nTR-Technician Repair\nUE-Under Evaluation\nUTR-Under Technician Repair\nW-Working\nWP-Waiting Parts\nParts Priced\nPending Internal Approval\nRSI-Repaired and Shipped Invoiced\nNEA-Need Evaluation Approved",
		// 	"default": "Submitted",
		// 	"width": "100"
		// }



		






	]
};

